### PR TITLE
Use improved CUDF JSON validation

### LINKED ISSUE #10457: [BUG] ScanJson and JsonToStructs allow unquoted control chars by default
**Describe the bug**
allowUnquotedControlChars is off by default in Spark for these expressions. But the CUDF parser happily allows them. There is no simple way to detect this without help from CUDF for some of the characters, because things like tab might show up as valid white space outside of a quoted string.

### LINKED ISSUE #10479: [BUG] JsonToStructs and ScanJson should return null for non-numeric, non-boolean non-quoted strings
**Describe the bug**
The way that from_json and the json scan work is that they will try to parse a number/boolean first and then if that works out the result is returned as a string.  This is also related to validation. If a single column is an invalid unquoted string, then then entire row needs to be invalidated.

In this case we are looking at unquoted values.  In JSON for boolean values only `true` and `false` are allowed. They are case sensitive so `TRUE` and `FALSE` are not valid.  Numbers have to look like the desired number type or they are not valid. 1.0 is not a valid int like with https://github.com/NVIDIA/spark-rapids/issues/10469.  Note that 1,000 is invalid in all cases for numbers, unless it is in a quoted string and is being read as a decimal value https://github.com/NVIDIA/spark-rapids/issues/10470.

Things get to be a little complicated because this is different for GetJsonObject or JsonTuple where everything that is valid is returned as a string. Note that I said is valid. `TRUE` is not a valid unquoted value, and it too would result in the entire line for GetJsonObject or JsonTuple being returned as null. 

I think to make this work we are either going to need some help from CUDF to have better validation. Or we are going to need complicates post processing by enabling CUDF to return quoted strings. I think the latter is going to give us the most flexibility, and then we can come back to CUDF and figure out how to make it work more effeciently.

### LINKED ISSUE #10534: [BUG] Need Improved JSON Validation
**Describe the bug**
In Spark if the JSON is not valid, then nothing from that document/row is parsed and all values are returned as null. CUDF validates some JSON, but not all of it the way that SPARK wants. So things like leading zeros on numbers can optionally invalidate an entire row. We will need CUDFs help to fully support config options like `allowNumericLeadingZeros`, `allowNonNumericNumbers`, `allowBackslashEscapingAnyCharacter`, and `allowUnquotedControlChars`.  All of these we kind of need in some form or another depending on the JSON operator.

### LINKED ISSUE #10911: from_json: when input is a bad json string, rapids would throw an exception.
from_json: when input is a bad json string, rapids would throw an exception.
## Repro:

CPU: 
```
scala> import org.apache.spark.sql.types.{IntegerType, StructType, StructField}

scala> val badJson = "\u0000\u0000\u0000A\u0001AAA"
badJson: String = ???A?AAA

scala> val df = Seq(badJson).toDS()
df: org.apache.spark.sql.Dataset[String] = [value: string]

scala> df.write.mode("OVERWRITE").parquet("TEMP")

scala> val df2 = spark.read.parquet("TEMP")
df2: org.apache.spark.sql.DataFrame = [value: string]

scala> val schema = StructType(StructField("a", IntegerType) :: Nil)
schema: org.apache.spark.sql.types.StructType = StructType(StructField(a,IntegerType,true))

scala> var parsed = df2.select(from_json($"value", schema))
parsed: org.apache.spark.sql.DataFrame = [from_json(value): struct<a: int>]

scala> parsed.show()
+----------------+
|from_json(value)|
+----------------+
|          {null}|
+----------------+
```

GPU:
```
$ $SPARK_HOME/bin/spark-shell --master local[*] --jars ${SPARK_RAPIDS_PLUGIN_JAR} 
--conf spark.plugins=com.nvidia.spark.SQLPlugin 
--conf spark.rapids.sql.enabled=true 
--conf spark.rapids.sql.explain=ALL --driver-java-options '-ea -Duser.timezone=UTC ' 
--conf spark.rapids.sql.expression.JsonTuple=true 
--conf spark.rapids.sql.expression.GetJsonObject=true 
--conf spark.rapids.sql.expression.JsonToStructs=true 
--conf spark.rapids.sql.expression.StructsToJson=true
```

```
scala> import org.apache.spark.sql.types.{IntegerType, StructType, StructField}
import org.apache.spark.sql.types.{IntegerType, StructType, StructField}

scala> val badJson = "\u0000\u0000\u0000A\u0001AAA"
badJson: String = ???A?AAA

scala> val df = Seq(badJson).toDS()
df: org.apache.spark.sql.Dataset[String] = [value: string]

scala> df.write.mode("OVERWRITE").parquet("TEMP")
24/05/28 02:01:29 WARN GpuOverrides: 
*Exec <DataWritingCommandExec> will run on GPU
  *Output <InsertIntoHadoopFsRelationCommand> will run on GPU
  ! <LocalTableScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.LocalTableScanExec
    @Expression <AttributeReference> value#16 could run on GPU


scala> val df2 = spark.read.parquet("TEMP")
df2: org.apache.spark.sql.DataFrame = [value: string]

scala> val schema = StructType(StructField("a", IntegerType) :: Nil)
schema: org.apache.spark.sql.types.StructType = StructType(StructField(a,IntegerType,true))

scala> var parsed = df2.select(from_json($"value", schema))
parsed: org.apache.spark.sql.DataFrame = [from_json(value): struct<a: int>]

scala> parsed.show()
24/05/28 02:01:31 WARN GpuOverrides: 
!Exec <CollectLimitExec> cannot run on GPU because the Exec CollectLimitExec has been disabled, and is disabled by default because Collect Limit replacement can be slower on the GPU, if huge number of rows in a batch it could help by limiting the number of rows transferred from GPU to CPU. Set spark.rapids.sql.exec.CollectLimitExec to true if you wish to enable it
  @Partitioning <SinglePartition$> could run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(from_json(StructField(a,IntegerType,true), value#20, Some(UTC)) as string) AS from_json(value)#26 will run on GPU
      *Expression <Cast> cast(from_json(StructField(a,IntegerType,true), value#20, Some(UTC)) as string) will run on GPU
        *Expression <JsonToStructs> from_json(StructField(a,IntegerType,true), value#20, Some(UTC)) will run on GPU
    *Exec <FileSourceScanExec> will run on GPU

24/05/28 02:01:31 ERROR Executor: Exception in task 0.0 in stage 7.0 (TID 7)
java.lang.IllegalStateException: No empty row count provided and the table read has no row count or columns
	at ai.rapids.cudf.Table.gatherJSONColumns(Table.java:1204)
	at ai.rapids.cudf.Table.readJSON(Table.java:1446)
	at ai.rapids.cudf.Table.readJSON(Table.java:1428)
	at org.apache.spark.sql.rapids.GpuJsonToStructs.$anonfun$doColumnar$2(GpuJsonToStructs.scala:179)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at org.apache.spark.sql.rapids.GpuJsonToStructs.$anonfun$doColumnar$1(GpuJsonToStructs.scala:177)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at org.apache.spark.sql.rapids.GpuJsonToStructs.doColumnar(GpuJsonToStructs.scala:175)
	at com.nvidia.spark.rapids.GpuUnaryExpression.doItColumnar(GpuExpressions.scala:250)
	at com.nvidia.spark.rapids.GpuUnaryExpression.$anonfun$columnarEval$1(GpuExpressions.scala:261)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at com.nvidia.spark.rapids.GpuUnaryExpression.columnarEval(GpuExpressions.scala:260)
	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:35)
	at com.nvidia.spark.rapids.GpuUnaryExpression.columnarEval(GpuExpressions.scala:260)
	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:35)
	at com.nvidia.spark.rapids.GpuAlias.columnarEval(namedExpressions.scala:110)
	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:35)
	at com.nvidia.spark.rapids.GpuProjectExec$.$anonfun$project$1(basicPhysicalOperators.scala:110)
	at com.nvidia.spark.rapids.RapidsPluginImplicits$MapsSafely.$anonfun$safeMap$1(implicits.s
```

### PR DESCRIPTION
**Pull request by revans2**
This depends on https://github.com/rapidsai/cudf/pull/15968 that was merged in recently

This fixes #10457 
This fixes #10479
This fixes #10534
This fixes #10911 

It also fixes some issue with JSON scan where if all of the input is invalid/empty lines before it would error out, but now our work around fixes that. It is not a final solution though.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala [hunk 1]
```diff
@@ -120,64 +120,41 @@ object GpuJsonReadCommon {
...
     }
+  }
 
-    withResource(cleaned) { _ =>
-      if (options.allowNonNumericNumbers) {
-        // Need to normalize the quotes to non-quoted to parse properly
-        withResource(ColumnVector.fromStrings(specialQuotedFloats: _*)) { quoted =>
-          withResource(ColumnVector.fromStrings(specialUnquotedFloats: _*)) { unquoted =>
-            cleaned.findAndReplaceAll(quoted, unquoted)
+  private def sanitizeInts(input: ColumnView, options: JSONOptions): ColumnVector = {
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala:135]: `options` is not used anymore.

**Review by abellina — COMMENTED** [review]
Looks good, I had one minor comment.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
@abellina please take another look

**Review by abellina — DISMISSED** [review]

**Review by ttnghia — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
@abellina please take another look. My changes to the text buffering accidentally filtered out too much for hive text files and our tests caught it. I made the change so it is now specific to JSON and added a test file to verify that the case when all of the lines are bogus or empty works as expected.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTextBasedPartitionReader.scala [hunk 2]
```diff
@@ -79,16 +79,24 @@ trait LineBuffererFactory[BUFF <: LineBufferer] {
...
+object FilterEmptyHostLineBuffererFactory extends LineBuffererFactory[HostLineBufferer] {
   override def createBufferer(estimatedSize: Long,
       lineSeparatorInRead: Array[Byte]): HostLineBufferer =
-    new HostLineBufferer(estimatedSize, lineSeparatorInRead)
+    new HostLineBufferer(estimatedSize, lineSeparatorInRead, true)
 }
 
 /**
  * Buffer the lines in a single HostMemoryBuffer with the separator inserted inbetween each of
  * the lines.
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTextBasedPartitionReader.scala:95]: nit, could you add a comment that HostLineBufferer is used for JSON and one for HostStringColBufferer that should be used for other formats?

**Review by abellina — APPROVED** [review]
just a nit.

**Pull request merged**